hibernate-rx-core/src/main/java/org/hibernate/rx/engine/impl/Cascade.java

```diff
@@ -97,6 +98,24 @@ public Cascade(final CascadingAction<C> action,
 	 * which is specific to each CascadingAction type
 	 */
 	public CompletionStage<Void> cascade() throws HibernateException {
+		return RxUtil.nullFuture().thenCompose( v -> {
+			CacheMode cacheMode = eventSource.getCacheMode();
+			if (action==CascadingActions.DELETE) {
+				eventSource.setCacheMode( CacheMode.GET );
+			}
+			final PersistenceContext persistenceContext = eventSource.getPersistenceContextInternal();
+			persistenceContext.incrementCascadeLevel();
+			try {
+				return cascadeInternal();
+			}
+			finally {
+				persistenceContext.decrementCascadeLevel();
```

Comment:
Is it OK ot decrement the cascade level even if cascadeInternal is not completed?

Reply:
Good question. I'm not sure. I *think* it's OK, and this was what we were doing previously, but I can't guarantee it was correct. Let me take a closer look.

Reply:
I think this could cause problems. I might be able to come up with a test case. Would that be worthwhile?

Reply:
@gbadner if I'm not mistaken the `cascadeLevel` is only used for checking that the user didn't do Bad things and throwing exceptions, so I don't think we invest time in writing a test for this.

On the other hand, I think I agree with you that&mdash;now that `cascadeInternal()` is doing more of its work asynchronously&mdash;this should be done in `handle()`. I'll make that change.

Reply:
Oh well, now I look more closely at this, the logic was just plain wrong. Sorry @DavideD, I didn't realize at first.

Fixed by 35bb3628692c7c7de93e2981879aa4ff8c623c1b